### PR TITLE
deliv label make package mandatory

### PR DESCRIPTION
**Pull request by bealdav**
TODO

- [x] put tracking in picking if only one package

Remarks:

because of no chatter in package I let label in picking

### CODE REVIEW

File: base_delivery_carrier_label/models/stock_picking.py [hunk 1]
```diff
@@ -87,21 +90,33 @@ def get_shipping_label_values(self, label):
...
-        """ Generate the labels.
-
-        A list of package ids can be given, in that case it will generate
-        the labels only of these packages.
+        """ Legacy method. Remove me after 12.0
+        """
+        raise UserError(
+            "Your delivery module depending on your carrier must call "
+            "action_generate_carrier_label() method "
+            "instead of generate_labels()")
```
Comment (yvaucher) [inline base_delivery_carrier_label/models/stock_picking.py:83]: What about a deprecation warning in the logs instead?

And call action_generate_carrier_label()

That way you can use the _logger defined after imports :)

File: base_delivery_carrier_label/models/stock_picking.py [hunk 2]
```diff
@@ -3,9 +3,12 @@
 # License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
 
 import base64
+import logging
 from odoo import _, api, fields, models
 from odoo.exceptions import UserError
 
+_logger = logging.getLogger(__name__)
```
Comment (yvaucher) [inline base_delivery_carrier_label/models/stock_picking.py:10]: Currently unused. (see comment about deprecation)

**Review by yvaucher — CHANGES_REQUESTED** [review]

**Comment by bealdav** [comment]
ping @yvaucher @florian-dacosta

File: base_delivery_carrier_label/views/stock.xml [hunk 3]
```diff
@@ -9,7 +9,8 @@
     <field name="arch" type="xml">
       <field name="state" position="before">
         <button name="action_generate_carrier_label" states="done"
-          string="Create Shipping Label" type="object"/>
+                help="Create Shipping Label 🚚"
```
Comment (bealdav) [inline base_delivery_carrier_label/views/stock.xml]:
```suggestion
                help="Create Shipping Label 🚚"
                invisible="[('carrier_code', '=', False)]"
```

**Review by yvaucher — APPROVED** [review]
LGTM

I'm currious of how 🚚 displays in Odoo ^^

**Comment by bealdav** [comment]
It's the job of the browser not odoo

hihi http://www.unicode.org/emoji/charts/full-emoji-list.html#1f69a

only one byte and a lot of semantic ...  

Allow to not waste space in UI #mobilefriendly

File: base_delivery_carrier_label/models/stock_picking.py [hunk 4]
```diff
@@ -85,23 +74,40 @@ def get_shipping_label_values(self, label):
...
+
+    def _set_a_default_package(self):
+        """ Pickings using this module must have a package
+            If not this method put it one silently
+        """
+        for picking in self:
+            move_lines = picking.move_line_ids.filtered(
+                lambda s: s.package_id and not s.result_package_id)
+            for move_line in move_lines:
+                move_line.write({'result_package_id': move_line.package_id.id})
```
Comment (florian-dacosta) [inline base_delivery_carrier_label/models/stock_picking.py]: I don't understand this part...
Why write result_package_id if a move line with package id is found?
Seems very strange to me to have a package_id and a result_package_id with the same package on the same move line...
And I don't get why you have to do this and no simply something like this : 
`move_lines = picking.move_line_ids.filtered( lambda s: not (s.package_id or s.result_package_id))
if move_lines:
    package = self.env['stock.quant.package'].create({})
    move_lines.write({'result_package_id': package.id})
`

**Review by florian-dacosta — COMMENTED** [review]
Just one remark because I don't get this part.

**Comment by florian-dacosta** [comment]
:+1:

**Review by angelmoya — APPROVED** [review]
:+1:

**Comment by bealdav** [comment]
This PR can be merged as it
but there is more here WIP
https://github.com/akretion/delivery-carrier/pull/125

**Comment by guewen** [comment]
> only one byte and a lot of semantic ...

4 bytes actually 😬

**Comment by bealdav** [comment]
ping Timo17100-c2c thanks for a review

**Comment by bealdav** [comment]
@yvaucher, please could you merge it ?

Thanks

**Comment by yvaucher** [comment]
@bealdav I don't have write access :-/

I'll do a take over of @Timo17100-c2c PR and add your refactoring.

**Comment by bealdav** [comment]
Thanks

**Comment by yvaucher** [comment]
Included it in https://github.com/OCA/delivery-carrier/pull/205

Regarding the number of approval here, we probably can merge the whole thing.

**Pull request closed without merging**